Added Attributes to node properties

kwave opened the pull request:
In order to releave @marvinroger a little bit, I've created a PR which contains the changes discussed in #12 . Let's proceed with the discussion directly in the PR

ThomDietrich commented:
Hey @kwave,
thanks for taking up this task!
Big coincidence, just this weekend I've developed an MQTT speaking daemon that uses "Homie Convention v2.1.0-alpha" 😃 and I was able to gain some head-on experience.

Not a complete review but two early addition:
* I believe you missed to add `homie/device-id/$nodes` to the list of "Device Properties"
* I realized that  "`$name` Friendly name" should also be avilable for properties
* The required flag should not be Yes as often as it is currently. The settable property should for example have a default false nature instead of requiring the topic. Other topics are similarly not needed with a proper default.

euphi commented on README.md:
```diff
@@ -176,32 +176,124 @@ Properties starting with a `$` are special properties. It must be one of the fol
+            <code>%</code> Percent<br>
+            <code>m</code> Meter<br>
+            <code>ft</code> Feet<br>
+            <code>#</code> Count or Amount
```
I miss a recommendation for (air) pressure. I propose Hectopascal (Unit `hPa`), because it is a (derived) SI-unit, other than Bar or mBar.  You may add "PSI" as imperial unit.

kwave replied:
I've added Pascal `Pa`(all other units are also listed in their in basic unit) and `psi`.

euphi commented on README.md:
````diff
@@ -176,32 +176,124 @@ Properties starting with a `$` are special properties. It must be one of the fol
+homie/ledstrip-device/ledstrip/led_3/$range → on,off
 homie/ledstrip-device/ledstrip/led_3 → on
 ```
 
````
The following (unchanged) section is inconsistent with the LED example.  I like the new definition with "On" or "Off" as state, so I propose to update the following section to make use of it. (No more "On"-state that is set to true or false).

kwave replied:
Why is it inconsistent? It describes how a client can set a property of the homie device, and how the device gives feedback about whether the value has been processed or not. Please clarify.

euphi replied:
The next section in README.MD is unchanged and thus somehow inconsistent:

```
Homie is state-based. You don't tell your smartlight to turn on, but you tell it to put it's on state to true. This especially fits well with MQTT, because of retained message.

For example, a kitchen-light device exposing a light node would subscribe to homie/kitchen-light/light/on/set and it would receive:

homie/kitchen-light/light/on/set ← true
```

According to your propasal it is better to use

`homie/kitchen-light/light/state/set ← on`

ThomDietrich replied:
agreed. Alternatively `homie/kitchen-light/light/power/set ← on`

marvinroger replied:
Agreed too. @kwave feel free to choose between `state` and `power`. 😉

kwave replied:
Ok done. Replaced "on" with power in all topics.

euphi commented on README.md:
```diff
@@ -176,32 +176,124 @@ Properties starting with a `$` are special properties. It must be one of the fol
+       <td>Device → Controller</td>
+       <td>Describes the format of data.</td>
+       <td><code>integer</code>, <code>float</code>, <code>boolean</code>, <code>string</code>, <code>enum</code> </td>
+       <td>Yes</td>
```
Format of boolean should be defined. (What is allowed, is it case sensitive?) "true" and "false" or also "TRUE", "FALSE", 0, 1?

kwave replied:
Good point. I've added this to the datatype.

kwave commented:
Fixed most of the stuff mentioned above.

> - I realized that "$name Friendly name" should also be available for Nodes
I've added this for nodes and also properties.
 

> - The required flag should not be Yes as often as it is currently. The settable property should for example have a default false nature instead of requiring the topic. Other topics are similarly not needed with a proper default.
I thought the same when writing down the changes. But there is one caveat that came to my mind: MQTT is message based. If you subscribe to ´property/#` you will receive a single message for each retained topic in **any** order. So how does the client know whether an attribute is not present (--> Use the default!) or if the retained message is about to be delivered later (--> Use the attribute value instead of the default). Any Ideas?

> - You've chosen a different regex format. Any reason why?
Your suggestion was lacking the possibility to add flags to the pattern. Thats all ;-)


> - In the README the convention structure is confusing. Maybe you could rename the section "Device properties" to "Device" and describe the deviceID topic as well as all homie/deviceID/$keyword topics under this section. The same goes for "Nodes" and for "Properties". As I mentioned in the original issue I also don't like the confusion around Property and properties. My suggestion would be that "Property" should only be used for the hierarchy level under Nodes (e.g. temperature) and the special .../$keyword topics should be addressed with "parameter", "meta-topic" or similar
I've changed the naming. Every subtopic starting with `$` is called "attribute" whereas the level below a node is still called "property"

ThomDietrich commented on README.md:
```diff
@@ -35,14 +35,17 @@ An ID MAY contain only lowercase letters from `a` to `z`, numbers from `0` to `9
 To efficiently parse messages, Homie defines a few rules related to topic names. The base topic you will see in the following convention will be `homie/`. You can however choose whatever base topic you want.
 
 * `homie` / **`device ID`**: this is the base topic of a device. Each device must have an unique device ID which adhere to the [ID format](#id-format).
```
"a unique"

kwave replied:
Fixed.

ThomDietrich commented on README.md:
```diff
@@ -18,7 +18,7 @@ You can find an implementation of the Homie convention:
 ## Background
 
-An instance of a physical piece of hardware (an Arduino, an ESP8266...) is called a **device**. A device has **device properties**, like the current local IP, the Wi-Fi signal, etc. A device can expose multiple **nodes**. For example, a weather device might expose a `temperature` node and an `humidity` node. A node can have multiple **node properties**. The `temperature` node might for example ex ...
+An instance of a physical piece of hardware (an Arduino, an ESP8266...) is called a **device**. A device has **attributes**, like the current local IP, the Wi-Fi signal, etc. A device can expose multiple **nodes**. For example, a weather device might expose a `temperature` node and an `humidity` node. A node can have multiple **properties**. The `temperature` node might for example expose a `degr ...
```
The example is not a good fit anymore. Unit is now it's dedicated attribute after all.
To illustrate the usage well we need more levels.

I'm not an expert on weather stations but how about something along this example:

* homie/
  * weatherstation/
    * base/
      * temperature
      * humidity
    * windsensor/
       * direction
       * speed
 * coffeemachine/
    * ...

...Plus all the additional attributes

marvinroger replied:
Good point.

kwave replied:
I've modified the example to be more precise. Basically added a new property for the battery level to the example.

ThomDietrich replied:
Be aware, that you missed out on the "base" and "windsensor" nodes. With the text as it is right now the meaning of nodes and properties doesn't get clear.

ThomDietrich commented on README.md:
```diff
@@ -35,14 +35,17 @@ An ID MAY contain only lowercase letters from `a` to `z`, numbers from `0` to `9
-### Device properties
 
-* `homie` / **`device ID`** / `$` **`device property`**: a topic starting with a `$` after the base topic of a device represents a device property. A device property MUST be one of these:
+### Device attributes
```
My suggestion was to move the sentence from above regarding the device below this headline and rename the headline to "Device". For the new reader it is pretty confusing that there is a headline e.g. "Node ...." and it immediately addresses attributes while the Node is not yet discussed. I'd suggest four big headlines "Homie Base", "Devices", "Nodes", "Properties".

marvinroger replied:
Agreed.

ThomDietrich replied:
Thanks but it's still not how I imagine it. Feel free to ask if my suggestion wasn't clear. Also I'm not saying that my suggestion is the only and best solution, so let me know if you disagree with anything ;)

ThomDietrich commented:
One general remark that might go beyond your PR: As you can see git management of long paragraphs is rather unfitting. Because of that it's often recommended to break down paragraphs into "one sentence per line" code. The markdown result will be the same but the git management of the document is better.

marvinroger commented on README.md:
```diff
@@ -4,7 +4,7 @@
 **Please note this v2 branch is a work-in-progress. It might change before the final release.**
 
-Version: **2.0.0**.
+Version: **2.1.0**.
```
As written in the above line, the v2 is not yet final, so we don't even need to bump the version. 👍

kwave replied:
I would suggest to version the smaller increments anyways. As the final version is still a v2, what about 2.0.1 for the changes discussed here?

marvinroger replied:
I would say no, but as the v2 is already widely used, let's keep it this way.

marvinroger commented on README.md:
```diff
@@ -139,6 +142,13 @@ To efficiently parse messages, Homie defines a few rules related to topic names.
+  <tr>
+    <td>$nodes</td>
+    <td>Device → Controller</td>
+    <td>Nodes the device exposes, with format `id` separated by a `,` if there are multiple nodes. For ranges, define the range after the `id`, within `[]` and separated by a `-`.</td>
```
Can you replace the `` ` `` with `<code>` tags? The `` ` `` doesn't work in HTML arrays.

kwave replied:
Strange, in my Atom-Markdown preview this works as expected. I've changed it.

euphi commented on README.md:
````diff
@@ -174,34 +183,147 @@ Properties starting with a `$` are special properties. It must be one of the fol
 ```
 
 A LED strip would look like this. Note that the topic for a range properties is the name of the property followed by a `_` and the index getting updated:
 
````
The term _range_ is now used in two different meanings in the convention.

1. To allow "arrays" of properties.
2. To define allowed value range

**--> possible Solution:** change `$range` to `$valuerange`
An alternative would be to change to range term (as used in the "old" convention) to something else, e.g. _vector_ .

kwave replied:
Agreed, this might get somehow confusing. I've changed the `$range` to `$format`, as "range" is the wrong term anyways. e.g. a regex is definitely not a "range", same for enums.

marvinroger replied:
`$format` seems perfect. @euphi ?

euphi commented on README.md:
```diff
@@ -174,34 +183,147 @@ Properties starting with a `$` are special properties. It must be one of the fol
+homie/ledstrip-device/ledstrip/led_1/$datatype → enum
+homie/ledstrip-device/ledstrip/led_1/$range → on,off
 homie/ledstrip-device/ledstrip/led_1 → on
+
```
Personally, I don't think it is a good idea to set the attribute of each property range-index individually. $settable, $unit, $datatype and $range (or $valuerange) should be the same for the complete range of the property and thus be defined on property level.
Ranges (or _vectors_) should simplify things and not make them more complicated. If you need individual attributes, use individual properties.

There may be a use for naming each ranged property:
 - I/O modules of industrial automation systems usually use formal names based on a simple numbering scheme (comparable to homie's _range_ (or _vector_)), but allow to set an optional symbolic name.

ThomDietrich commented:
Hey guys, great progress so far!
I did a little brainstorming with a colleague today and we found room for two improvements I'd like to discuss with you:

1. **Property class/category:** A property is a distinct reading or characteristic value produced by a node of a device. The other side of the network, the discovering entity, needs to know about the nature of this value (e.g. to show a thermometer in a GUI). We are already providing attributes to characterize its unit, range, pretty name and "set'ability" but not a clear type.
  A not-settable property with the unit "°C" does at best *imply* that the property is a temperature reading of a thermometer, not e.g. a thermostat or a nominal temperature. The discovering entity should not have to resort to guessing and interpreting the other attributes.
  I propose a **counted number of property classes** to characterize the nature of a property. The `$class` or `$category` attribute is required but can be "Unspecified"/"Other"/"" for special non-trivial properties.

2. **Conversion Attribute:** When working with embedded systems or low-energy micro controllers, we can not expect the device to be capable to convert a raw sensor reading into a final human-comprehendible value. An example would be a temperature sensor reading between [0..255] which would map onto a degree Celcius value with decimal places.
  I propose the introduction of a `$conversion` property attribute. The published String value is either "x"/"" for no conversion or the right hand of an `f(x)` formula.

Both points are of course open for discussion. What do you think?

marvinroger commented:
@ThomDietrich I see a problem with *property* classes. Imagine you have an `heater` node, with two properties: `mode` (`off,comfort,night`), and `temperature`. How do you assing a "class" to `mode`, for example? The most logical thing to do would be to assing classes to *nodes*, not *properties*. Therefore we would have `heater`, `light`, `weatherstation`... But this would introduce another problem: for the UI to be able to show a thermometer with the temperature embedded in the picture, we would have to define a standard set of properties for each class. Which defeat the flexibility of the homie convention, right?

For the conversion attribute, I am not that sure. It's rather complicated, and I don't see in practice what kind of value can't be computed directly on-device. I mean, even a 3$ chip is able to read a temperature in `°C`, `°F`.

marvinroger commented:
@kwave 

> Your suggestion was lacking the possibility to add flags to the pattern. Thats all ;-)

A simple regex pattern is enough, there's no need to use any flags in such a case, right? (global, unicode, multiline...). The only exception being the *ignore flag*, but for ease of implementation by parser and simplicity, I think it's better to do `[a-zA-Z]` rather than `/[a-z]/i`. Objections?

euphi commented:
I didn't had the time yet to look at @ThomDietrich 's proposal, but I suggest to move this discussion to an own issue, because it is out of scope of the original intention of this PR, so this PR can be merged soon.

ThomDietrich commented:
Agreed. Let's merge the PR soon and incorporate further additions later.
There are many details I'd like to add or incorporate and individual PRs for these, based on the state of this PR now would be easier to manage...
